### PR TITLE
fix: stash working tree should work when everything/nothing is staged

### PR DESCRIPTION
**Pull request by golden-expiriensu**
Stash working tree is running 3 commands so there is some edge cases which we should take into account:
1. If everything is staged don't do anything
2. If nothing is staged don't try to save index in the stash and then later pop it (either errors or pops non-related stashes)

Addresses: https://github.com/altsem/gitu/pull/100#issuecomment-2025832482

### CODE REVIEW

**Comment by altsem** [comment]
Trying to do some digging if there's a better way to handle all this.
There's multiple places where things might fail. What if there's an ongoing merge etc?

If only git had a `git stash push --worktree` :man_facepalming:.
The `git stash pop 1` is scary because it assumes nothing went wrong.

One idea is to return an error code from run_cmd and react to it.

I found this function in the git source code:
https://github.com/git/git/blob/d6fd04375f9196f8b203d442f235bd96a1a068cc/builtin/stash.c#L1288

**Comment by altsem** [comment]
The C function does something like this I believe (123 would be the pid):
```bash
$ git diff --name-only | GIT_INDEX_FILE=.git/index.stash.123 git update-index --ignore-skip-worktree-entries --add --remove --stdin
```
And then it writes a "stash commit" somehow after.
Although doing this results in not just the diffed file being stage.

```bash
$ GIT_INDEX_FILE=.git/index.stash.123 git status
On branch fix/stash-working-tree-edge-cases
Your branch is up to date with 'golden-expiriensu/fix/stash-working-tree-edge-cases'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	deleted:    .cliffignore
	deleted:    .envrc
	deleted:    .github/dependabot.yml
	deleted:    .github/workflows/cachix.yml
...
```

Git creates this `.git/index.stash.123` file temporarily.

**Comment by altsem** [comment]
Maybe just changing the `state.run_cmd()` function to actually return an error if the command failed is the way to go. :thinking: 
That way we'd just break execution in the `stash_worktree` fn.

Would not have to modify anything of what you already wrote in this case I think.

File: src/ops/stash.rs [hunk 1]
```diff
@@ -33,25 +34,31 @@ impl OpTrait for StashWorktree {
     fn get_action(&self, _target: Option<&TargetData>) -> Option<Action> {
         let update_fn = move |state: &mut State, term: &mut Term| -> Res<()> {
             if state.prompt.state.status().is_done() {
+                if is_working_tree_empty(&state.repo)? {
```
Comment (altsem) [inline src/ops/stash.rs:37]: There's no need to initiate a prompt when `is_working_tree_empty`.
See `state.prompt.set` below at line 70

**Review by altsem — CHANGES_REQUESTED** [review]

File: src/ops/stash.rs (hunk 1 above)
Comment (golden-expiriensu) [reply src/ops/stash.rs:37]: Yeah, that makes sense. Do we want to show something to user? Like "Cannot stash: working tree is empty"

Comment (altsem) [reply src/ops/stash.rs]: Sounds like a good idea!

File: src/tests/mod.rs [hunk 2]
```diff
@@ -487,21 +487,44 @@ mod stash {
...
-                    key('i'),
-                    key_code(KeyCode::Enter),
-                    key('z'),
-                    key('p'),
-                    key('1'),
+                    key('w'),
+                    key('t'),
+                    key('e'),
+                    key('s'),
+                    key('t'),
```
Comment (altsem) [inline src/tests/mod.rs:527]: I'd love to solve the syntax for sending these events somehow.
Would need one for keybind config anyway.
Something like, "jjjuzwtest<enter>".
And perhaps a test helper fn to select lines:
`select("Unstaged changes.*");`

Doesn't belong in this PR though.

**Review by altsem — APPROVED** [review]

File: src/ops/stash.rs (hunk 1 above)
Comment (golden-expiriensu) [reply src/ops/stash.rs]: I have encountered very strange situation.., can you try to run current tests and after that test things locally? Locally I am getting expected error message when working tree is empty, but there is no message in test. Did I miss something?

File: src/tests/mod.rs (hunk 2 above)
Comment (golden-expiriensu) [reply src/tests/mod.rs:527]: Wow, this sounds awesome! Implementing "jjjuzwtest<enter>" doesn't sound very hard tho. Selecting on the other hand looks pretty hard in my head, since we have everything inside the terminal. Does it have something like `read_to_string`? I need to research it a little bit

Comment (altsem) [reply src/tests/mod.rs:527]: It's all in an in-memory buffer for tests. Seems a bit hard to work with.

File: src/ops/stash.rs (hunk 1 above)
Comment (altsem) [reply src/ops/stash.rs]: The test tries to `zw` and then input `test` and Enter.
It breaks after the change with the prompt I noticed. Pressing keys while the error is up will close it :) Line 500

File: src/ops/stash.rs [hunk 3]
```diff
@@ -36,39 +41,88 @@ impl OpTrait for StashWorktree {
...
-            move |state: &mut State, _term: &mut Term| -> Res<()> {
-                state.prompt.set(PromptData {
+            move |state: &mut State, term: &mut Term| -> Res<()> {
+                if is_working_tree_empty(&state.repo)? {
+                    state
+                        .current_cmd_log_entries
+                        .push(Arc::new(RwLock::new(CmdLogEntry::Error(
+                            "Cannot stash: working tree is empty".to_string(),
+                        ))));
+                    state.prompt.reset(term)?;
```
Comment (altsem) [inline src/ops/stash.rs:76]:
```suggestion
```
Should work without! There's no prompt open here yet.

**Review by altsem — CHANGES_REQUESTED** [review]

File: src/ops/stash.rs [hunk 4]
```diff
@@ -36,39 +41,88 @@ impl OpTrait for StashWorktree {
...
+                    return Ok(());
+                }
+                Ok(state.prompt.set(PromptData {
                     prompt_text: "Name of the stash:".into(),
                     update_fn: Rc::new(update_fn),
-                });
-                Ok(())
+                }))
```
Comment (altsem) [inline src/ops/stash.rs:82]: Clippy is complaining about this in the build i noticed.
The `ci.sh` script has some goodies i run locally myself.

File: src/ops/stash.rs (hunk 1 above)
Comment (golden-expiriensu) [reply src/ops/stash.rs]: My bad 🤦‍♂️ Thanks for taking a look!

File: src/ops/stash.rs (hunk 3 above)
Comment (golden-expiriensu) [reply src/ops/stash.rs]: Indeed, done

File: src/ops/stash.rs (hunk 4 above)
Comment (golden-expiriensu) [reply src/ops/stash.rs]: Yeah, I was running only `cargo test`, now I will run ci script. Very strange rule to be honest, this should be delegated to compiler :)
Done

**Review by altsem — APPROVED** [review]

**Comment by altsem** [comment]
Looked good! :)
